Add Light Unit Slider (Light Intensity, Temperature, Exposure)

johnpars opened the pull request:
### Purpose of this PR
This PR introduces the [proposed improvement](https://www.figma.com/file/Jb2OfbkPvAjlz9RbyMhnvf/HDRP-UI-IMPROVEMENTS?node-id=0%3A1) to UI/UX for light units. It adds a new slider UI drawer that loads a table of helpful value ranges, providing more context to the user about light unit values; through icons, markers, and tooltips. This will help the user to more easily reason about the variety of physically based values in HDRP, when authoring lighting and exposure.

## Light Intensity Units (EV, Nits, Lumen, Candela, Lux)
![LightIntensity](https://user-images.githubusercontent.com/28882975/93141603-7d98d100-f6b2-11ea-832f-5932b29da27a.gif)

## Temperature Units (Kelvin)
![Temperature](https://user-images.githubusercontent.com/28882975/93141613-812c5800-f6b2-11ea-8a78-2c31c6e5c8e9.gif)

## Exposure Units (EV)
![Exposure](https://user-images.githubusercontent.com/28882975/93141616-838eb200-f6b2-11ea-87bf-58b85e1cbd66.gif)

## Editor Theme
![EditorThemes](https://user-images.githubusercontent.com/28882975/93161073-c61ab380-f6df-11ea-907e-80f784394ce6.gif)

## Slider is constant across all punctual light intensity units and spot reflector
![ConstantSlider](https://user-images.githubusercontent.com/28882975/93594559-1154f080-f984-11ea-832d-ff9eb29efcd3.gif)

---
### Testing status

- [ ] Checked new UI names with UX convention
- [x] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)

---
### Comments to reviewers

- This PR implements the new UI drawer for **Light Intensities**, **Temperature**, and **Exposure** (volume component), but has not yet been integrated to the **scene view override exposure**, or the **physically based sky exposure**.
- The slider for light intensities and exposure is modeled as a piece-wise function. At the cost of not being continuous and not monotonically increasing, it has the benefit of easily being able to accommodate the wide value distributions for certain light units, The PR does include an exponential version (which fits a curve to a low, mid, and high value), but its currently unused and can be removed. I have sent an email to our artists to gather additional feedback on this piece-wise distribution.
- For icon loading, currently they are not loaded with the HDRP editor resource loading mechanism (reloader). This is because we need to take special care to load icons so that they are skinned properly for the editor and it's color theme. Maybe it's possible to add Reloader support for icon case?

iM0ve reviewed (changes requested):
**Review status:** In Progress

**Issues:**

1) **FIXED** - Can not input precise values. Same happens when upgrading a project, light values are changed.

![values](https://user-images.githubusercontent.com/36502659/92893486-60e66a00-f422-11ea-9856-f00d5dd387d7.gif)

2) **FIXED** - Overcast sky starts at 80k in the editor. In our own doc or wikipedia, we can see the value stated is 1000-2000 lux. Im not sure if this is an issue with terminology or with the values.

<img width="406" alt="Screen Shot 2020-09-11 at 11 52 12" src="https://user-images.githubusercontent.com/36502659/92896686-4497fc80-f425-11ea-820a-e6c64c0672d6.png">

sources:
https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@10.0/manual/Physical-Light-Units.html
https://en.wikipedia.org/wiki/Daylight

3) Suggestion to change the tooltip for this one to something like "Exceeds realistic sunlight - 120K Lux". (no dot before dash). I think having static warning about 120k would benefit users more than just showing their own value in the tooltip, this way they can quickly evaluate how much they are overshooting.

<img width="406" alt="Screen Shot 2020-09-11 at 11 59 34" src="https://user-images.githubusercontent.com/36502659/92898321-8ecdad80-f426-11ea-8ab1-8fccfd5ac7cc.png">

4) **FIXED** - Concern. Im not sure about this one, but maybe the icons are a bit too transparent? I noticed the sun slider icon right away but somehow I didnt even see the Temperature one. Compared to all the other editor icons these are hard to spot. Matter of discussion.

<img width="406" alt="Screen Shot 2020-09-11 at 12 22 59" src="https://user-images.githubusercontent.com/36502659/92902199-beca8000-f429-11ea-96c8-855c340241bf.png">

5) Temperatures about sunlight are not too useful on Point/Spot/Area lights. We could show typical light bulb/candle values instead?

<img width="406" alt="Screen Shot 2020-09-11 at 12 26 57" src="https://user-images.githubusercontent.com/36502659/92902780-3b5d5e80-f42a-11ea-91d9-dceb0b5c38ed.png">

6) Undo no longer works. You can now also break rendering if you switch to EV100 and try to undo

![break](https://user-images.githubusercontent.com/36502659/92908203-add03d80-f42e-11ea-8d09-28c13a182df0.gif)

7) I think the new Light Appearance mode might cause workflow issues. We might want to not implement the feature. [More details ](https://unity.slack.com/archives/C01A9B2JFNK/p1600248431067800)

8) When a value is converted to a different type, the icon indicates that it changes the lighting intensity. When in reality after conversion the value should be exactly the same.

![bm1fqTEJZo](https://user-images.githubusercontent.com/36502659/93321398-c8355d00-f81a-11ea-8337-390910212ece.gif)


**Whats tested:**

1) Tooltips
2) Upgrading a project
3) UX
4) Functional testing

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightUI.cs:
```diff
@@ -59,6 +59,8 @@ enum AdvancedMode
 
         readonly static ExpandedState<Expandable, Light> k_ExpandedState = new ExpandedState<Expandable, Light>(~(-1), "HDRP");
 
+        readonly static LightUnitSliderUIDrawer s_LightUnitSliderUIDrawer = new LightUnitSliderUIDrawer();
```
It should be `k_LightUnitSliderUIDrawer` instead

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+
+namespace UnityEditor.Rendering.HighDefinition
+{
+    class LightUnitSlider
```
A small comment here describing the purpose of the class would be nice here.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+            public const float k_ThumbTooltipSize   = 10;
+        }
+
+        protected LightUnitSliderUIDescriptor m_Descriptor;
```
The descriptor is not meant to be modified, so you should add a `readonly`  keyword for safety.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+            EditorGUI.LabelField(thumbMarkerRect, GetLightUnitTooltip(tooltip, value));
+        }
+
+        GUIContent GetLightUnitTooltip(string baseTooltip, float value)
```
This sounds like a useful utility here. Maybe you can extract a static function to convert a value to a readable human string (float value) -> string.
And then use it here.

Also, maybe we can use existing API like https://github.com/Humanizr/Humanizer#metric-numerals or similar?

johnpars replied:
Hey @fredericv-unity3d, due to artist request we now just display the whole value here instead of abbreviating it into shorter terms. So I will mark resolved for now

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+            value.floatValue = GUI.HorizontalSlider(rect, value.floatValue, range.x, range.y, GUI.skin.horizontalSlider, GUI.skin.horizontalSliderThumb);
+        }
+
+        static float Remap(float v, float x0, float y0, float x1 = 0f, float y1 = 1f) => x1 + (v - x0) * (y1 - x1) / (y0 - x0);
```
A short comment about the kind of remapping occurs here can be nice.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+        }
+    }
+
+    class ExponentialLightUnitSlider : LightUnitSlider
```
A small comment here describing the purpose of the class would be nice here.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+        }
+    }
+
+    class TemperatureSlider : LightUnitSlider
```
A small comment here describing the purpose of the class would be nice here.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+            return new GUIContent(string.Empty, tooltip);
+        }
+
+        protected virtual void DoSlider(Rect rect, SerializedProperty value, Vector2 range)
```
A small comment on protected API can be nice to know what is expected in the implementation.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -0,0 +1,316 @@
+
+        static float Remap(float v, float x0, float y0, float x1 = 0f, float y1 = 1f) => x1 + (v - x0) * (y1 - x1) / (y0 - x0);
+
+        protected virtual float GetPositionOnSlider(float value)
```
A small comment on protected API can be nice to know what is expected in the implementation.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSliderSettings.cs:
```diff
@@ -0,0 +1,208 @@
+    static class LightUnitSliderDescriptors
+    {
+        // Lux
+        public static LightUnitSliderUIDescriptor k_LuxDescriptor = new LightUnitSliderUIDescriptor(
```
`public static` don't follow the `k_` standard, you can directly use pascal case here and for other instances.

fredericv-unity3d reviewed (changes requested):
I noticed a few minor things that needs changes and a few code standard issues.
But the overall PR is quite good.

Can you make the changes?

johnpars commented:
Thanks @fredericv-unity3d - I have implemented almost all your notes barring the unit value humanizing. 

Note also that I am currently finishing an implementation of a piece-wise slider (due to the exponential slider currently not being fit well enough for the Lux value distribution). Additionally, I refactored the icon resource loading mechanism to support the different editor themes (HDRP's resource loading mechanism didn't make this easy). Due to those changes I will ping you for a re-review if that is alright.

@iM0ve - Note that at the time of your review the icon rendering indeed was not complete, but has since should now be correct for both editor themes. For the rest of your notes, I will investigate & follow up ASAP tomorrow.

johnpars commented:
Hi @iM0ve, an update for you based on my recent commits:

> 1) Can not input precise values. Same happens when upgrading a project, light values are changed.

This should now be fixed. 

> 2) Overcast sky starts at 80k in the editor. In our own doc or wikipedia, we can see the value stated is 1000-2000 lux. Im not sure if this is an issue with terminology or with the values.

Awaiting input from UX on this one (cc @MaddalenaUnity). In the PR notes I linked to the original icon/value sheet which I derived all these values from.

> 3) Suggestion to change the tooltip for this one to something like "Exceeds realistic sunlight - 120K Lux". (no dot before dash).

Also waiting for UX thoughts on this one, I do agree though that it would be a positive change.

> 4) Concern. Im not sure about this one, but maybe the icons are a bit too transparent? I noticed the sun slider icon right away but somehow I didnt even see the Temperature one. Compared to all the other editor icons these are hard to spot. Matter of discussion.

Should now be fixed for both editor themes.

> 5) Temperatures about sunlight are not too useful on Point/Spot/Area lights. We could show typical light bulb/candle values instead?

Similar to 2, not fixed as I do not have the information/resources (icons) to improve these current values for punctual lights

> 6) Undo no longer works. You can now also break rendering if you switch to EV100 and try to undo

I believe this should now be fixed (both the issues with Undo as well as the NaN in EV units).

johnpars commented:
Note: more discussion on this can be found on **#ux-lighting-units-improvements**

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -148,26 +165,47 @@ GUIContent GetLightUnitTooltip(string baseTooltip, float value)
+                formatValue = value.ToString("#0.0");
 
-            string tooltip = baseTooltip + " - " + formatValue + " " + m_Descriptor.unitName;
+            var tooltip = baseTooltip + " | " + formatValue + " " + unit;
```
This is mostly for your information, but you can use the [string interpolation syntax](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) like this: `$"{baseTooltip}|{formatValue} {unit}"`

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -202,19 +240,141 @@ protected override float GetPositionOnSlider(float value)
+        // Piecewise function indexed by value ranges.
+        private readonly Dictionary<int, Piece> m_PiecewiseFunctionMap = new Dictionary<int, Piece>();
+
+        Func<float, float> GetTransformation(float x0, float x1, float y0, float y1)
```
`GetTransformation` is not exactly what is does.

I would use instead `NewTransformation` or `CreateTransformation` because it instantiate a lambda function (thus creating GC allocation).

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -202,19 +240,141 @@ protected override float GetPositionOnSlider(float value)
+                piece.transform = GetTransformation(x0, x1, y0, y1);
+
+                // Compute the inverse
+                CoreUtils.Swap(ref x0, ref y0);
```
Note: swap can be done like this: `(x0, y0) = (y0, x0)`

johnpars replied:
Here I went with your other suggestion to just swap directly in the call

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -202,19 +240,141 @@ protected override float GetPositionOnSlider(float value)
+                // Compute the inverse
+                CoreUtils.Swap(ref x0, ref y0);
+                CoreUtils.Swap(ref x1, ref y1);
+                piece.inverseTransform = GetTransformation(x0, x1, y0, y1);
```
IMO swapping and then calling the method is a bit overkill here.

Just calling `GetTransformation(y0, y1, x0, x1)` is enough as the swapped values are not reused for something else.
I find it easier to understand the intent here.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -202,19 +240,141 @@ protected override float GetPositionOnSlider(float value)
+        {
+            public Vector2 domain;
+            public Vector2 range;
+            public Func<float, float> transform;
```
As the algorithm does not change between transform and inverse transform, the use of `Func` is not appropriate: it is a reference function, possibly a null pointer.

You can store the `m` and `b` variables for direct and inverse transform and have a method to apply the algorithm instead.

If you are looking for reusable code you can have:
```
static float DoTransform(in float x, in float m, in float b) => (m * x) + b;
float Transform(in float x) => DoTransform(x, directM, directB);
float InverseTransform(in float x) => DoTransform(x, inverseM, inverseB);
```

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -202,19 +240,141 @@ protected override float GetPositionOnSlider(float value)
+        }
+
+        // Piecewise function indexed by value ranges.
+        private readonly Dictionary<int, Piece> m_PiecewiseFunctionMap = new Dictionary<int, Piece>();
```
The key type `int` is disturbing here. You can directly use `Vector2` with the Dictionary and it will probably void a few bugs.

Dictionary will use the `GetHashCode()` method to bin the values and will use the `IEquatable<Self>.Equals(Self other)` method if `IEquatable<Self>` is implemented, otherwise it will use `object.Equals`.

So it means that if you use directly `Vector2` as Key type, you will achieve the same binning and still have `O(1)` time access, but it will handle equality cases when the `GetHashCode()` values are the same for different `Vector2`  (hash collision).

So other point is about understanding the code, `Vector2` is more explicit then `int` to indicate a range.

Can you please make this change?

fredericv-unity3d replied:
If I want to push further, I would use a new type pattern enclosing the range:
```
struct ValueRange: IEquatable<ValueRange>
{
    Vector2 m_Range;

    public ValueRange(in Vector2 value) => m_Range = value;
    public Vector2 ToVector2() => m_Range;
    // Insert generated Equality methods by ReSharper here
    
    public static explicit operator Vector2(in ValueRange v) => v.m_Range;
    public static explicit operator ValueRange(in Vector2 v) => new ValueRange(v);
}
```

Pros and cons of the newtype pattern:
Pros:
* The compiler will detect invalid uses of a Vector2 as a ValueRange and inversely. (Type check error)
* It is easier to review
* It helps a lot separating code in small understable and standalone pieces.

Cons:
* In C# it involves a lot of code.

I don't request the use of the newtype pattern as it is quite verbose, it is for your information you may need this pattern in the future.

(But I do request the change to `Vector2` at least :) )

johnpars replied:
Nice, thank you for all this info! Adding it to the toolbelt :) In this case, if it's alright, I have switched to using the `Vector2 `directly as the key. If I have extra time before end of week I may try to add this pattern, but currently need to allocate the remaining time to addressing the various QA and Artist request for the PR)

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -202,19 +240,141 @@ protected override float GetPositionOnSlider(float value)
+
+                if (x >= p.domain.x && x <= p.domain.y)
+                {
+                    piece = p;
```
You should return a boolean to indicate if the search was successful.

fredericv-unity3d reviewed (changes requested):
Thanks for resolving the issue and doing a clean PR 👍 .

I have still a few requests concerning the added code, can you make the changes?

(To be explicit: I also added potential changes at some points, but may be overkill for those use cases and I don't request to apply them, these are only for information)

iM0ve reviewed (changes requested):
Updated status for existing issues. Added 7&8 
See my Review comment above

johnpars commented:
Hi @iM0ve, thanks for the extra review. 

To address your notes:
- **3: Fixed**
- **7: Fixed** after discussion with artists in thread: the filter/temperaturewill be the default.
- **8: Fixed**, however there is an ongoing discussion with lighting artist on how this behavior should work when spot reflector is enabled.
- **5: Not Fixed**, this is a more UI/UX note which I will forward to @MaddalenaUnity

johnpars commented:
Hi @fredericv-unity3d, thanks also for your review.

Your notes should all be addressed (note: for your note regarding Humanizing tooltip value, there is ongoing discussion with artists if it should be abbreviated or not).

If it's alright, I will request another re-review. I just pushed one last significant change to the slider that makes it always operate internally in Lumen. This was done to directly fix @iM0ve feedback number 8. 

It removed a lot of code (in the settings), added a slider specifically for converting between light units, and it elevated the access level of `ConvertLightIntensity` in `HDLightUI` to internal.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -261,24 +261,27 @@ struct Piece
+        private readonly Dictionary<Vector2, Piece> m_PiecewiseFunctionMap = new Dictionary<Vector2, Piece>();
 
-        Func<float, float> GetTransformation(float x0, float x1, float y0, float y1)
+        static void NewTransformation(float x0, float x1, float y0, float y1, out float m, out float b)
```
Now, the naming is still incorrect for other reason. `NewTransformation` was a good choice when it instantiated a lambda, now it would be more `ComputeTransformationParameters`.

I think that I had a confusing request in the previous review, the `NewTransformation` naming was good with the lambda pattern, but now we use directly an algorithm with parameters, the naming should be different.

Sorry about that.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -302,28 +305,24 @@ public PiecewiseLightUnitSlider(LightUnitSliderUIDescriptor descriptor) : base(d
         }
 
-        void UpdatePiece(ref Piece piece, float x)
+        bool UpdatePiece(ref Piece piece, float x)
```
Maybe indicate a small comment about the semantic of the returned boolean and that it performs a piece search based on `x` value.

It is not very obvious what kind of update is expected to be performed.

fredericv-unity3d commented on com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:
```diff
@@ -372,6 +373,62 @@ protected override void DoSlider(Rect rect, SerializedProperty value, Vector2 sl
+        private LightUnit m_Unit;
+
+        // Note: these should be in sync with LightUnit
+        private static string[] m_UnitNames =
```
`static readonly string[] k_UnitNames = `

fredericv-unity3d reviewed (changes requested):
I found very minor issues that should be changed, can you update them?

fredericv-unity3d commented:
You can request as many reviews as you need, it is always helpful!
Thanks for your work!

iM0ve commented:
No new issues found. Updated the review. Only issue 5 still remains. Looks good. I will wait for all the changes to land and do one last check before approving.

johnpars commented:
Thanks @fredericv-unity3d - I implemented your requests, and also added one last improvement that should satisfy the request of the artist (basically, need to satisfy the case of spot reflector; where the slider should be unchanged between light units). There was an unfortunate limitation to the `LightUtil` conversion that required me manually recompute some things (more info in the comment). Instead of improving the `LightUtil ` I decided to accomplish it all in the slider, as I was nervous about changing light unit conversion at this point since it might actually impact things on the runtime.

johnpars commented:
Hi @iM0ve, for issue 5, it's certainly a doable change but I haven't been able to confirm alignment on it with UX. Additionally, I assume the change would also require new icons which I don't think are currently on hand from icon artist. If the change to the tooltip is enough however, it should be quickly doable.

Note also: I pushed a change to the light intensity slider that should be an improvement for how it behaves with spot reflector. Its probably the area of most concern in your next round of testing.

fredericv-unity3d reviewed (approved):
This is good for me!

johnpars commented:
Note: This PR now fully implements the originally designed spec. However, we have the following feedback from artists:
- Improvement of iconography for temperature slider (specifically, waiting for two new icons - info is updated in the icon google sheet. Note that adding this should address the remaining issue flagged by @iM0ve).
- Add a right-click context menu on the slider, offering a dropdown of presets.

sebastienlagarde commented:
Hey, so I will merge the current PR and we will do the improvement in a following one. THanks

iM0ve commented:
Still wanted to do a final pass before merge :/ please warn next time.
Did a quick check on reflector and Undo. Found that undo doesnt work with reflector properly and switching between certain light value types. For example works if Candela->Lumen->Undo and doesnt work Candela->Lux->Undo


![type](https://user-images.githubusercontent.com/36502659/93865334-d4179800-fcce-11ea-9ecd-0cb4820570e8.gif)
![refl](https://user-images.githubusercontent.com/36502659/93865340-d548c500-fcce-11ea-9f46-dcbeccf4845f.gif)

sebastienlagarde commented:
arf sorry about that. was guessing it was done. Good catch
